AutoPas/AutoPas
# Refactor Iterator Tests

# Description

- [x] Change `IteratorBehavior` `enum` to an `Option` class.
- [x] Refactor `RegionParticleIteratorTest`
- [x] Merge common functionality of `IteratorTest`, `RegionParticleIteratorTest`, and `ParticleIteratorTest`
- [x] Add Test for Iterator copying (copy constructor, copy assignment)
- [x] Throw out `TouchableParticle`
- [x] Rename `IteratorTest` to `ParticleIteratorInterfaceTest`
- [x] Discourage particle insertion while iterating (see below)

Currently, we don't explicitly say that we do not support particle addition while iterating. However, I strongly suggest discouraging this! While most containers (except LCR) correctly handle the case that a particle is inserted in a cell that the iterator of the same thread is currently covering or will cover in the future, insertion by thread A into a cell that is currently iterated by thread B is unsafe! This could be circumvented if all containers work with insertion buffers, or even better when the logic handler buffers particle insertion.

## Related Pull Requests

- merges into #579 

## Resolved Issues

- [x] closes #536 -> not supported and checked in the logic handler

# State of the tests before this PR

## IteratorTest

* uses the AutoPas interface
* all parametrized for container, csf, const, priorForceCalc

* ParticleAdditionAndIteratorTestNormal
    * parametrized for container, csf, const, priorForceCalc
    * testAdditionAndIteration
        * Add some strategically placed owned (3) and halo (7) particles around critical spots of the domain
        * Do or don't do a force calculation
        * Iterate with normal iterator and every behavior and check the number of found particles
        * Iterate domain+halo with region iterator and every behavior
        * checkRegionIteratorForAllParticles for every behavior
            * use regular Iterator over container
            * start a region iter around every particle and check dereferenced iters for equivalence
            * check that every particle is only found once
        * Move particles by skin/2
        * checkRegionIteratorForAllParticles for every behavior

* RangeBasedIterator
    * parametrized for container, csf, const, priorForceCalc
    * testRangeBasedIterator
        * Add some strategically placed owned (3) and halo (7) particles around critical spots of the domain
        * Do or don't do a force calculation
        * iterates all particles with range based and iterator based loop and compares what is found
        * NOTE: Sketchy! Comparison what was found should be done better

* testOpenMP${Region}Iterators${Behavior}
    * parametrized for container, csf, const, priorForceCalc
    * testOpenMPIterators
        * Add 500 random particles
        * Add 50 random halo particles
        * Do or don't do a force calculation
        * Iterate with OpenMP and touch everything according to behavior
        * Again without OpenMP and check that everything was touched once
    
* RegionIteratorDeletion
    * parametrized for container, csf, const, priorForceCalc
    * testRegionIteratorDeletion
        * fill AutoPas object with 200 random owned particles (might be in halo)
        * Do or don't do a force calculation
        * Save all generated particles in a vector
        * Use region iterator to delete all particles from a region and save what was deleted
        * Use regular Iterator to check everything that should be there is there
        * NOTE: looks improveable


## ParticleIteratorTest

* All tests use OpenMP if available

* testFullIterator_EFEFFEEFEF
    * parametrized for numThreads, numAdditionalParticleVectors
    * testAllParticlesFoundPattern
        * Create vector of cells, some filled with N particles according to pattern
        * Create some additional particle vectors
        * Use ParticleIterator loop, collect all found IDs
        * Check if all expected IDs were found

* testFullIterator_FEFEEFFEFE
    * same as above but different pattern

* testFullIterator_deletion
    * Create vector of cells, some filled with N particles according to pattern
    * Use iterator to loop over cells and delete all particles
    * Use iterator to loop over cells and confirm that there is nothing left

* testFullIterator_mutable
    * Create vector of cells, some filled with N particles according to pattern
    * Use iterator over cells to modify velocity data
    * Use iterator over cells to confirm modification

* testIteratorBehavior${Container}
    * Create container
    * Add two particles, one inner and one in halo
    * For every behavior check that the respective particles are found by the iterator


## RegionParticleIteratorTest

* uses container objects

* test${Container}RegionParticleIterator${Behavior}${Threads}
    * fill a container with 100 particles
    * touch them with a region iterator 
    * check if everything that is supposed to be touched is touched

* test${Container}RegionParticleIteratorEmpty
    * create an empty LC container
    * call iterator loop
    * check that the iterator is invalid

* testLinkedCellsRegionParticleIteratorCopyConstructor
    * fill a LC container with 100 particles
    * create two iterators where iterator2 is created by a copy constructor
    * touch all particles with iterator2
    * iterate with iterator1 and check that everything was touched

* testLinkedCellsRegionParticleIteratorCopyAssignment
    * same as above but with copy assignment instead of copy constructor

* testLinkedCellsRegionParticleIteratorSparseDomain
    * strategically place some particles in the domain
    * mark which should be touched and which not
    * iterate container and check that expectations match via inBox()
